mixer/pkg/protobuf/yaml/dynamic/builder.go

```diff
@@ -0,0 +1,280 @@
+		compiler    Compiler
+		skipUnknown bool
+	}
+)
+
+// NewEncoderBuilder creates an EncoderBuilder.
+func NewEncoderBuilder(resolver yaml.Resolver, compiler Compiler, skipUnknown bool) *Builder {
+	return &Builder{
+		resolver:    resolver,
+		compiler:    compiler,
+		skipUnknown: skipUnknown}
+}
+
+// Build builds a DynamicEncoder
+func (c Builder) Build(msgName string, data map[interface{}]interface{}) (Encoder, error) {
+	m := c.resolver.ResolveMessage(msgName)
+	if m == nil {
+		return nil, fmt.Errorf("cannot resolve message '%s'", msgName)
+	}
+
+	return c.buildMessage(m, data, true)
+}
+
+func (c Builder) buildMessage(md *descriptor.DescriptorProto, data map[interface{}]interface{}, skipEncodeLength bool) (Encoder, error) {
+	var err error
+	var ok bool
+
+	me := messageEncoder{
+		skipEncodeLength: skipEncodeLength,
+	}
+	isMap := md.GetOptions().GetMapEntry()
+
+	keys := make([]string, 0, len(data))
+	for kk := range data {
+		var k string
+		if k, ok = kk.(string); !ok {
+			return nil, fmt.Errorf("error processing message '%s':%v got %T want string", md.GetName(), kk, kk)
+		}
+		keys = append(keys, k)
+	}
+	// Ensure consistent log output.
+	sort.Strings(keys)
+
+	for _, k := range keys {
+		v := data[k]
+		fd := yaml.FindFieldByName(md, k)
+		if fd == nil {
+			if c.skipUnknown {
+				continue
+			}
+			return nil, fmt.Errorf("field '%s' not found in message '%s'", k, md.GetName())
+		}
+
+		switch fd.GetType() {
+		case descriptor.FieldDescriptorProto_TYPE_STRING:
+			var ma []interface{}
+			if fd.IsRepeated() {
+				if ma, ok = v.([]interface{}); !ok {
+					return nil, fmt.Errorf("unable to process: %v, got %T, want: []interface{}", fd, v)
+				}
+			} else {
+				ma = []interface{}{v}
+			}
+
+			for _, vv := range ma {
+				var fld *field
+				if fld, err = c.buildPrimitiveField(vv, fd, isMap && k == "key"); err != nil {
+					return nil, fmt.Errorf("unable to encode field: %v. %v", fd, err)
+				}
+				me.fields = append(me.fields, fld)
+			}
+
+			// primitives
+		case descriptor.FieldDescriptorProto_TYPE_DOUBLE,
+			descriptor.FieldDescriptorProto_TYPE_FLOAT,
+			descriptor.FieldDescriptorProto_TYPE_INT64,
+			descriptor.FieldDescriptorProto_TYPE_UINT64,
+			descriptor.FieldDescriptorProto_TYPE_INT32,
+			descriptor.FieldDescriptorProto_TYPE_FIXED64,
+			descriptor.FieldDescriptorProto_TYPE_FIXED32,
+			descriptor.FieldDescriptorProto_TYPE_BOOL,
+			descriptor.FieldDescriptorProto_TYPE_UINT32,
+			descriptor.FieldDescriptorProto_TYPE_SFIXED32,
+			descriptor.FieldDescriptorProto_TYPE_SFIXED64,
+			descriptor.FieldDescriptorProto_TYPE_SINT32,
+			descriptor.FieldDescriptorProto_TYPE_SINT64,
+			descriptor.FieldDescriptorProto_TYPE_ENUM:
+
+			packed := fd.IsPacked() || fd.IsPacked3()
+			if fd.IsRepeated() && !packed {
+				return nil, fmt.Errorf("unpacked primitives not supported: %v", fd)
+			}
+
+			var fld *field
+			if fld, err = c.buildPrimitiveField(v, fd, isMap && k == "key"); err != nil {
```

Comment:
It seems you cannot have a non string key in json. So, this code would change a bit

Reply:
update to name to `noExpr`